include/dmlc/optional.h

```diff
@@ -0,0 +1,170 @@
+/*!
+ * Copyright (c) 2016 by Contributors
+ * \file optional.h
+ * \brief Container to hold optional data.
+ */
+#ifndef DMLC_OPTIONAL_H_
+#define DMLC_OPTIONAL_H_
+
+#include <iostream>
+#include <utility>
+#include <algorithm>
+
+#include "./base.h"
+#include "./logging.h"
+#include "./type_traits.h"
+
+namespace dmlc {
+/* \brief dummy type for assign null to optional */
+struct nullopt_t {
+  constexpr nullopt_t(int) {}
+};
+
+// Assign null to optional: optional<T> x = nullopt;
+constexpr nullopt_t nullopt(0);
+
+/*!
+ * \brief c++17 compatible optional class.
+ *
+ * At any time an optional<T> instance either
+ * hold no value (string representation "None")
+ * or hold a value of type T.
+ */
+template<typename T>
+class optional {
+ public:
+  /*! \brief construct an optional object that contains no value */
+  optional() : is_none(true) {}
+  /*! \brief construct an optional object with value */
+  explicit optional(const T& value) {
+    is_none = false;
+    *reinterpret_cast<T*>(&val) = value;
```

Comment:
This is wrong when value have non-trivial destructor (destructor will get invoked on val, which is not T). Use the following code instead
```c++
new ((&val)  T(value)
```